src/consensus/fork_choice.zig

```diff
@@ -0,0 +1,1871 @@
+
+            // Don't count children currently marked as invalid
+            // child must exist in tree
+            if (!(self.isCandidate(&child) orelse return error.MissingChild)) {
+                continue;
+            }
+
+            if (child_weight > maybe_best_child_weight or
+                (maybe_best_child_weight == child_weight and
+                child.order(maybe_best_child.*) == .lt))
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /// Checks if `deepest_child` is the deepest among its siblings.
+    ///
+    /// - A node is the deepest if no sibling has:
+    ///   1. A greater height.
+    ///   2. The same height but higher stake weight.
+    ///   3. The same height & stake but a higher slot number.
+    ///
+    /// - If `deepest_child` has no parent, it is the root and deepest by default.
+    fn isDeepestChild(self: *HeaviestSubtreeForkChoice, deepest_child: *const SlotAndHash) bool {
+        const maybe_deepest_child_weight =
+            self.stakeVotedSubtree(deepest_child) orelse return false;
+        const maybe_deepest_child_height = self.getHeight(deepest_child) orelse return false;
+        const maybe_parent = self.getParent(deepest_child);
+
+        // If there's no parent, this must be the root
+        if (maybe_parent == null) {
+            return true;
+        }
+        // Saftety: maybe_parent cannot be null due to the if check above.
+        const parent = maybe_parent.?;
+        // Get the other chidren of the parent. i.e. siblings of the deepest_child.
+        var children = self.getChildren(&parent) orelse return false;
+
+        for (children.keys()) |child| {
+            const child_height = self.getHeight(&child) orelse return false;
+            const child_weight = self.stakeVotedSubtree(&child) orelse return false;
+
+            const height_cmp = std.math.order(child_height, maybe_deepest_child_height);
+            const weight_cmp = std.math.order(child_weight, maybe_deepest_child_weight);
+            const slot_cmp = std.math.order(child.slot, deepest_child.slot);
+
+            switch (height_cmp) {
+                .gt => return false,
+                .eq => switch (weight_cmp) {
+                    .gt => return false,
+                    .eq => switch (slot_cmp) {
+                        .lt => return false,
+                        else => {},
+                    },
+                    else => {},
+                },
+                else => {},
+            }
+        }
+
+        return true;
+    }
+
+    fn getParent(
+        self: *const HeaviestSubtreeForkChoice,
+        slot_hash_key: *const SlotAndHash,
+    ) ?SlotAndHash {
+        if (self.fork_infos.get(slot_hash_key.*)) |fork_info| {
+            return fork_info.parent;
+        }
+        return null;
+    }
+
+    // TODO: Change this to return an iterator.
```

Comment:
can we create a ticket for this -- might be even worth adding it to this pr if its not too big

Reply:
Added https://github.com/Syndica/sig/issues/556